hotel/src/crypto/aes.rs

```diff
@@ -0,0 +1,155 @@
+use super::keymgr::{KEYMGR0_REGS, Registers};
+
+pub struct AesEngine {
+    regs: *mut Registers,
+    client: Cell<Option<&'static AesClient>>,
+}
+
+impl AesEngine {
+    const unsafe fn new(regs: *mut Registers) -> AesEngine {
+        AesEngine {
+            regs: regs,
+            client: Cell::new(None),
+        }
+    }
+
+    pub fn set_client(&self, client: &'static AesClient) {
+        self.client.set(Some(client));
+    }
+
+    pub fn setup(&self, key_size: aes::KeySize, key: &[u8]) -> Result<isize, SyscallError> {
+        let ref regs = unsafe { &*self.regs }.aes;
+
+        self.enable_all_interrupts();
+        regs.ctrl.set(regs.ctrl.get() | key_size as u32 | AesModule::Enable as u32);
+        let mut i = 0;
+        for word in key.chunks(4) {
+            if i >= 8 {
+                println!("Input key too long!");
+                return Err(SyscallError::InvalidState);
+            }
+            let d = word.iter()
+                .map(|b| *b as u32)
+                .enumerate()
+                .fold(0, |accm, (i, byte)| accm | (byte << (i * 8)));
+            regs.key[i].set(d);
+            i += 1;
+        }
+        regs.key_start.set(1);
+
+        Ok(0)
+    }
+
+    pub fn set_encrypt_mode(&self, encrypt: bool) -> Result<isize, SyscallError> {
+        let ref regs = unsafe { &*self.regs }.aes;
+
+        let flag = aes::Mode::Encrypt as u32;
+        if encrypt {
+            regs.ctrl.set(regs.ctrl.get() | flag);
+        } else if regs.ctrl.get() & flag != 0 {
+            regs.ctrl.set(regs.ctrl.get() ^ flag);
+        }
+
+        Ok(0)
+    }
+
+    pub fn crypt(&self, input: &[u8]) -> Result<isize, SyscallError> {
+        let ref regs = unsafe { &*self.regs }.aes;
+
+        let mut written_bytes = 0;
+        let mut written_words = 0;
+        for word in input.chunks(4) {
+            if regs.wfifo_full.get() != 0 || written_bytes >= 16 {
+                break;
+            }
+            let d = word.iter()
+                .map(|b| *b as u32)
+                .enumerate()
+                .fold(0, |accm, (i, byte)| accm | (byte << (i * 8)));
+            regs.wfifo_data.set(d);
+            written_bytes += word.len() as isize;
+            written_words += 1;
+        }
+
+        // Make sure we wrote 128 bits (4 words)
+        for _ in written_words..4 {
+            regs.wfifo_data.set(0);
+        }
+
+        Ok(written_bytes)
+    }
+
+    pub fn read_data(&self, output: &mut [u8]) -> Result<usize, SyscallError> {
+        let ref regs = unsafe { &*self.regs }.aes;
+
+        let mut i = 0;
+        while regs.rfifo_empty.get() == 0 {
+            if output.len() > i * 4 + 3 {
+                let word = regs.rfifo_data.get();
+                output[i * 4 + 0] = (word >> 0) as u8;
+                output[i * 4 + 1] = (word >> 8) as u8;
+                output[i * 4 + 2] = (word >> 16) as u8;
+                output[i * 4 + 3] = (word >> 24) as u8;
+                i += 1;
```

Comment:
Just increment `i` by 4 here instead, you multiply it by 4 everywhere it's used.